implementations/rust/ockam/ockam_api/tests/auth.rs

```diff
@@ -0,0 +1,104 @@
+use ockam_core::Result;
+use ockam_identity::TrustEveryonePolicy;
+use ockam_node::Context;
+
+#[ockam_macros::test]
+async fn credential(ctx: &mut Context) -> Result<()> {
+    // Create a secure channel API listener:
+    {
+        let v = Vault::create();
+        let a = Identity::create(ctx, &v).await?;
+        a.create_secure_channel_listener("api", TrustEveryonePolicy, &InMemoryStorage::new())
+            .await?;
+    }
+
+    // Create the authority signer service:
+    {
+        let v = Vault::create();
+        let a = Identity::create(ctx, &v).await?;
+        let s = signer::Server::new(a, InMemoryStorage::new());
+        ctx.start_worker("signer", s).await?;
+    }
+
+    // Create an admin that can manage the authenticator service:
+    let admin = {
+        let v = Vault::create();
+        Identity::create(ctx, &v).await?
+    };
+
+    // Create the authority authenticator service:
+    {
+        let e_store = InMemoryStorage::new();
+        let m_store = InMemoryStorage::new();
+        let mut auth = direct::Server::new(m_store, e_store, mk_signer(ctx).await?);
+        auth.set_admin(&admin.identifier()?);
+        ctx.start_worker("auth", auth).await?;
+    }
+
+    // Create an enroller and have the admin add it as authorised to add members:
+    let enroller = {
+        let v = Vault::create();
+        let e = Identity::create(ctx, &v).await?;
+        let a2a = admin
+            .create_secure_channel("api", TrustEveryonePolicy, &InMemoryStorage::new())
+            .await?;
+        let mut c = direct::Client::new(route![a2a, "auth"], ctx).await?;
+        c.add_enroller(IdentityId::new(e.identifier()?.key_id()))
+            .await?;
+        e
+    };
+
+    // Get the authority's signer key:
+    let mut signer_client = mk_signer(ctx).await?;
+    let auth_identity = signer_client.identity().await?;
+
+    // Connect to the API channel from the enroller:
+    let e2a = enroller
+        .create_secure_channel("api", TrustEveryonePolicy, &InMemoryStorage::new())
+        .await?;
+
+    // Create a member identity:
+    let v = Vault::create();
+    let m = Identity::create(ctx, &v).await?;
+
+    // Add the member via the enroller's connection:
+    let mut c = direct::Client::new(route![e2a, "auth"], ctx).await?;
+    c.add_member(IdentityId::new(m.identifier()?.key_id()))
+        .await?;
+
+    // Open a secure channel from member to authenticator:
+    let m2a = m
+        .create_secure_channel("api", TrustEveryonePolicy, &InMemoryStorage::new())
+        .await?;
+
+    let mut c = direct::Client::new(route![m2a, "auth"], ctx).await?;
+
+    // Get a fresh member credential and verify its validity:
+    let cred = c.credential().await?.to_owned();
+    let atts = c.validate(&cred).await?;
+    assert!(atts.get("id").is_some());
+    assert!(atts.get("ts").is_some());
+
+    // Here we pretend to be on a different node and want to validate the credential:
+    {
+        let v = Vault::create();
```

Comment:
I'm not saying it must be added to that test, but noting for now: it's critically important that credential is obtained from the other party through a secure channel AND it's checked that credential subject is the same as secure channel participant. I'm worried that this will be often forgotten (which breaks the whole system), so that maybe we should make verifier take subject id as an extra argument so that it will perform the check as well?